5_make_offxmls/make_offxml.py

```diff
@@ -0,0 +1,548 @@
+from loguru import logger
+from datasets import load_from_disk
+
+from qcportal import PortalClient
+
+from openff.toolkit import ForceField
+
+sys.path.insert(0, str(pathlib.Path(__file__).parent.parent))
+from from_finlay.molecular_classes import (  # noqa: E402
+    MMComponent,
+    Bond,
+    Angle,
+)
+from from_finlay import process_SMIRKS as ffps  # noqa: E402
+from from_finlay import process_mmcomponents as ffpmm  # noqa: E402
+from from_finlay.coverage import check_all_components_fully_covered_parallel_chunks  # noqa: E402
+
+logger.remove()
+
+
+def load_specificity_config(config_file: str | pathlib.Path) -> dict:
+    """
+    Load specificity configuration from JSON file.
+
+    Parameters
+    ----------
+    config_file : str | pathlib.Path
+        Path to JSON configuration file.
+
+    Returns
+    -------
+    dict
+        Configuration with *_specificities for keys defining the specificity for bonds, angles, etc..
+
+    Examples
+    --------
+    >>> config = load_specificity_config("config.json")
+    >>> "bond_specificities" in config
+    True
+    """
+    logger.info(f"Configuration file contents from {config_file}:")
+    with open(config_file, "r") as f:
+        content = f.read()
+        logger.info(content)
+    with open(config_file, "r") as f:
+        config = json.load(f)
+
+    return config
+
+
+def summarize_all_types(
+    mm_component_types: dict[int, dict[str, list[MMComponent]]],
+) -> None:
+    """
+    Log statistical summary of component types across specificity levels.
+
+    Parameters
+    ----------
+    mm_component_types : dict[int, dict[str, list[MMComponent]]]
+        Component organization: {specificity_level: {smirks_pattern: [components]}}.
+
+    Examples
+    --------
+    >>> types = {0: {"[#6:1]-[#6:2]": [bond1, bond2]}}
+    >>> summarize_all_types(types)
+    # Logs: "Total unique component types at specificity level 0: 1"
+    """
+    for specificity_num, components_by_type in mm_component_types.items():
+        component_type_counts = ffpmm.get_mm_component_type_num(components_by_type)
+        logger.info(
+            f"Total unique component types at specificity level {specificity_num}: {len(component_type_counts)}"
+        )
+        logger.info("Most common component types:")
+        for smirks, count in sorted(
+            component_type_counts.items(), key=lambda item: item[1], reverse=True
+        )[:10]:
+            logger.info(f"{smirks}: {count}")
+
+    # Flatten the component types for overall statistics
+    all_component_type_counts = ffpmm.flatten_mm_component_types(mm_component_types)
+    logger.info(
+        f"% of component types with count < 5: {sum(1 for count in all_component_type_counts.values() if count < 5) / len(all_component_type_counts) * 100:.2f}%"
+    )
+    logger.info(
+        f"% of component types with count = 1: {sum(1 for count in all_component_type_counts.values() if count == 1) / len(all_component_type_counts) * 100:.2f}%"
```

Comment:
(not blocking) These are a bit over the cusp of "this is too much logic to put on one line" for me. These could be difficult to debug/understand.